test/jdk/java/lang/ref/ReferenceRefersTo.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright (c) 2020, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @summary Basic functional test of Reference.refersTo.
+ */
+
+import java.lang.ref.Reference;
+import java.lang.ref.ReferenceQueue;
+import java.lang.ref.PhantomReference;
+import java.lang.ref.SoftReference;
+import java.lang.ref.WeakReference;
+
+public class ReferenceRefersTo {
+    private static final class TestObject {
+        public final int value;
+
+        public TestObject(int value) {
+            this.value = value;
+        }
+    }
+
+    private static final void fail(String msg) throws Exception {
+        throw new RuntimeException(msg);
+    }
+
+    private static final <T extends Reference>
+    void test(T ref,
+              TestObject expectedValue,
+              TestObject unexpectedValue,
+              String kind) throws Exception {
+        if ((expectedValue != null) && ref.refersTo(null)) {
+            fail(kind + "refers to null");
+        } else if (!ref.refersTo(expectedValue)) {
+            fail(kind + " doesn't refer to expected value");
+        } else if (ref.refersTo(unexpectedValue)) {
+            fail(kind + " refers to unexpected value");
```

Comment:
A possible failure mode is to implement the predicate incorrectly, perhaps flipping the sense of the predicate. This may help notice such. It's distinct from the null test, since null is a special value that might be handled differently.